// lineupengine/build/src/cell/internal/QuadTreeNode.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/** @internal */
// eslint-disable-next-line max-classes-per-file
var AQuadTreeNode = /** @class */ (function () {
    function AQuadTreeNode(index, rowFirst, rowLast, colFirst, colLast, rowTotal, colTotal) {
        this.index = index;
        this.rowFirst = rowFirst;
        this.rowLast = rowLast;
        this.colFirst = colFirst;
        this.colLast = colLast;
        this.rowTotal = rowTotal;
        this.colTotal = colTotal;
        this.parent = null;
    }
    Object.defineProperty(AQuadTreeNode.prototype, "rowCount", {
        get: function () {
            // since inclusive
            return this.rowLast - this.rowFirst + 1;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AQuadTreeNode.prototype, "colCount", {
        get: function () {
            return this.colLast - this.colFirst + 1;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AQuadTreeNode.prototype, "width", {
        get: function () {
            return this.rowTotal;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AQuadTreeNode.prototype, "height", {
        get: function () {
            return this.colTotal;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AQuadTreeNode.prototype, "rowId", {
        get: function () {
            var id = this.index < 2 ? '0' : '1';
            var p = this.parent;
            while (p !== null) {
                id = "".concat(p.index < 2 ? '0' : '1').concat(id);
                p = p.parent;
            }
            return id;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AQuadTreeNode.prototype, "colId", {
        get: function () {
            var id = this.index % 2 === 0 ? '0' : '1';
            var p = this.parent;
            while (p !== null) {
                id = "".concat(p.index % 2 === 0 ? '0' : '1').concat(id);
                p = p.parent;
            }
            return id;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AQuadTreeNode.prototype, "id", {
        get: function () {
            var id = String(this.index);
            var p = this.parent;
            while (p !== null) {
                id = "".concat(p.index, "-").concat(id);
                p = p.parent;
            }
            return id;
        },
        enumerable: false,
        configurable: true
    });
    return AQuadTreeNode;
}());
/** @internal */
var QuadTreeLeafNode = /** @class */ (function (_super) {
    __extends(QuadTreeLeafNode, _super);
    function QuadTreeLeafNode() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'leaf';
        return _this;
    }
    return QuadTreeLeafNode;
}(AQuadTreeNode));
export { QuadTreeLeafNode };
/**
 * @internal
 */
export var TOP_LEFT = 0;
/**
 * @internal
 */
export var TOP_RIGHT = 1;
/**
 * @internal
 */
export var BOTTOM_LEFT = 2;
/**
 * @internal
 */
export var BOTTOM_RIGHT = 3;
/** @internal */
var QuadTreeInnerNode = /** @class */ (function (_super) {
    __extends(QuadTreeInnerNode, _super);
    function QuadTreeInnerNode() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'inner';
        /**
         * 0 | 1
         * 2 | 3
         */
        _this.children = [];
        return _this;
    }
    Object.defineProperty(QuadTreeInnerNode.prototype, "colMiddle", {
        get: function () {
            return Math.floor(this.colFirst + this.colCount / 2) - 1;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(QuadTreeInnerNode.prototype, "rowMiddle", {
        get: function () {
            return Math.floor(this.rowFirst + this.rowCount / 2) - 1;
        },
        enumerable: false,
        configurable: true
    });
    return QuadTreeInnerNode;
}(AQuadTreeNode));
export { QuadTreeInnerNode };
