// lineupengine/build/src/cell/ACellRenderer.js
import { range } from '../logic';
import { clear } from '../internal/index';
import { BOTTOM_LEFT, BOTTOM_RIGHT, QuadTreeInnerNode, QuadTreeLeafNode, TOP_LEFT, TOP_RIGHT, } from './internal/QuadTreeNode';
var template = "<header></header>\n<aside></aside>\n<main><div></div></main><style></style>";
var leafCount = 4; // don't split further than 4x4 grids
var ACellRenderer = /** @class */ (function () {
    function ACellRenderer(root) {
        this.root = root;
        this.poolLeaves = [];
        this.poolInner = [];
        // private readonly fragment: DocumentFragment;
        /** @internal */
        this.tree = null;
        // eslint-disable-next-line no-param-reassign
        root.innerHTML = template;
        root.classList.add('lineup-cell-engine');
        // this.fragment = root.ownerDocument!.createDocumentFragment();
    }
    Object.defineProperty(ACellRenderer.prototype, "doc", {
        get: function () {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            return this.root.ownerDocument;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ACellRenderer.prototype, "body", {
        get: function () {
            return this.root.children[2];
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ACellRenderer.prototype, "colHeader", {
        get: function () {
            return this.root.firstElementChild;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ACellRenderer.prototype, "rowHeader", {
        get: function () {
            return this.root.children[1];
        },
        enumerable: false,
        configurable: true
    });
    ACellRenderer.prototype.init = function () {
        var _this = this;
        var body = this.body;
        var rowHeader = this.rowHeader;
        var colHeader = this.colHeader;
        var oldTop = body.scrollTop;
        var oldLeft = body.scrollLeft;
        body.addEventListener('scroll', function () {
            var left = body.scrollLeft;
            var top = body.scrollTop;
            if (oldTop === top && oldLeft === left) {
                return;
            }
            var isGoingDown = top > oldTop;
            var isGoingRight = left > oldLeft;
            oldTop = top;
            oldLeft = left;
            rowHeader.scrollTop = top;
            colHeader.scrollLeft = left;
            _this.onScroll(left, top, body.clientWidth, body.clientHeight, isGoingDown, isGoingRight);
        });
        this.recreate();
    };
    ACellRenderer.sliceHeight = function (ctx, start, end) {
        var height = (end - start + 1) * ctx.defaultRowHeight;
        for (var _i = 0, _a = ctx.exceptions; _i < _a.length; _i++) {
            var ex = _a[_i];
            if (ex.index < start) {
                continue;
            }
            if (ex.index > end) {
                break;
            }
            height += ex.height - ctx.defaultRowHeight; // change to exception height
        }
        return height;
    };
    ACellRenderer.buildTree = function (row, col) {
        var build = function (index, rowFirst, rowLast, colFirst, colLast, rowTotal, colTotal) {
            var rowCount = rowLast - rowFirst + 1;
            var colCount = colLast - colFirst + 1;
            if (rowCount <= leafCount && colCount <= leafCount) {
                return new QuadTreeLeafNode(index, rowFirst, rowLast, colFirst, colLast, rowTotal, colTotal);
            }
            var inner = new QuadTreeInnerNode(index, rowFirst, rowLast, colFirst, colLast, rowTotal, colTotal);
            var leftSlice = ACellRenderer.sliceHeight(col, colFirst, inner.colMiddle);
            var rightSlice = colTotal - leftSlice;
            var topSlice = ACellRenderer.sliceHeight(row, rowFirst, inner.rowMiddle);
            var bottomSlice = rowTotal - topSlice;
            inner.children.push(build(TOP_LEFT, rowFirst, inner.rowMiddle, colFirst, inner.colMiddle, topSlice, leftSlice));
            inner.children.push(build(TOP_RIGHT, rowFirst, inner.rowMiddle, inner.colMiddle + 1, colLast, topSlice, rightSlice));
            inner.children.push(build(BOTTOM_LEFT, inner.rowMiddle + 1, rowLast, colFirst, inner.colMiddle, bottomSlice, leftSlice));
            inner.children.push(build(BOTTOM_RIGHT, inner.rowMiddle + 1, rowLast, inner.colMiddle + 1, colLast, bottomSlice, rightSlice));
            inner.children.forEach(function (c) {
                // eslint-disable-next-line no-param-reassign
                c.parent = inner;
            });
            return inner;
        };
        return build(0, 0, row.numberOfRows - 1, 0, col.numberOfRows - 1, row.totalHeight, col.totalHeight);
    };
    ACellRenderer.prototype.recreate = function () {
        var _this = this;
        var context = this.privateContext;
        var body = this.body;
        this.tree = ACellRenderer.buildTree(context.row, context.col);
        // clear
        var root = body.firstElementChild;
        Array.from(root.children).forEach(function (c) { return _this.recycle(c); });
        this.clearPool();
        var col = range(body.scrollLeft, body.clientWidth, context.col.defaultRowHeight, context.col.exceptions, context.col.numberOfRows);
        var row = range(body.scrollTop, body.clientHeight, context.row.defaultRowHeight, context.row.exceptions, context.row.numberOfRows);
        root.dataset.node = this.tree.type;
        root.dataset.id = this.tree.id;
        root.style.width = "".concat(this.tree.width, "px");
        root.style.height = "".concat(this.tree.height, "px");
        this.render(this.tree, root, row.first, row.last, col.first, col.last);
    };
    ACellRenderer.prototype.onScroll = function (left, top, width, height, _isGoingDown, _isGoingRight) {
        var context = this.privateContext;
        var col = range(left, width, context.col.defaultRowHeight, context.col.exceptions, context.col.numberOfRows);
        var row = range(top, height, context.row.defaultRowHeight, context.row.exceptions, context.row.numberOfRows);
        var root = this.body.firstElementChild;
        if (this.tree) {
            this.render(this.tree, root, row.first, row.last, col.first, col.last);
        }
    };
    ACellRenderer.prototype.renderLeaf = function (leaf, parent) {
        var doc = this.doc;
        var children = Array.from(parent.children);
        // eslint-disable-next-line no-param-reassign
        parent.dataset.leafCols = String(leaf.colCount);
        if (children.length > 0) {
            clear(parent);
        }
        for (var row = leaf.rowFirst; row <= leaf.rowLast; row += 1) {
            for (var col = leaf.colFirst; col <= leaf.colLast; col += 1) {
                var item = children.shift();
                if (item != null) {
                    var change = this.updateCell(item, row, col);
                    if (change && change !== item) {
                        children.unshift(item);
                        item = change;
                    }
                }
                else {
                    item = this.createCell(doc, row, col);
                }
                item.dataset.row = String(row);
                item.dataset.col = String(col);
                parent.appendChild(item);
            }
            parent.appendChild(doc.createElement('br'));
        }
        return parent;
    };
    ACellRenderer.prototype.render = function (node, parent, rowFirst, rowLast, colFirst, colLast) {
        var _this = this;
        if (node.type === 'leaf') {
            return this.renderLeaf(node, parent);
        }
        var inner = node;
        var create = function (index) {
            var _a, _b;
            var child = inner.children[index];
            var newNode;
            if (child.type === 'inner') {
                newNode = (_a = _this.poolInner.pop()) !== null && _a !== void 0 ? _a : _this.doc.createElement('div');
            }
            else {
                newNode = (_b = _this.poolLeaves.pop()) !== null && _b !== void 0 ? _b : _this.doc.createElement('div');
            }
            newNode.dataset.node = child.type;
            newNode.dataset.id = child.id;
            return _this.render(child, newNode, rowFirst, rowLast, colFirst, colLast);
        };
        var placeholder = function (index) {
            var _a;
            var child = inner.children[index];
            var placeHolderNode = (_a = _this.poolInner.pop()) !== null && _a !== void 0 ? _a : _this.doc.createElement('div');
            placeHolderNode.dataset.node = 'placeholder';
            placeHolderNode.dataset.id = child.id;
            placeHolderNode.style.width = "".concat(child.width, "px");
            placeHolderNode.style.height = "".concat(child.height, "px");
            return placeHolderNode;
        };
        var children = Array.from(parent.children);
        var showLeft = !(inner.colFirst > colLast || inner.colMiddle < colFirst);
        var showRight = !(inner.colMiddle > colLast || inner.colLast < colFirst);
        var showTop = !(inner.rowFirst > rowLast || inner.rowMiddle < rowFirst);
        var showBottom = !(inner.rowMiddle > rowLast || inner.rowLast < rowFirst);
        if (children.length === 0) {
            parent.appendChild(showLeft && showTop ? create(TOP_LEFT) : placeholder(TOP_LEFT));
            parent.appendChild(showRight && showTop ? create(TOP_RIGHT) : placeholder(TOP_RIGHT));
            parent.appendChild(this.doc.createElement('br'));
            parent.appendChild(showLeft && showBottom ? create(BOTTOM_LEFT) : placeholder(BOTTOM_LEFT));
            parent.appendChild(showRight && showBottom ? create(BOTTOM_RIGHT) : placeholder(BOTTOM_RIGHT));
            return parent;
        }
        // can reuse
        {
            var topLeftNode = children[TOP_LEFT];
            var down = showLeft && showTop;
            if (down !== (topLeftNode.dataset.node !== 'placeholder')) {
                // no match
                parent.replaceChild(down ? create(TOP_LEFT) : placeholder(TOP_LEFT), topLeftNode);
                this.recycle(topLeftNode);
            }
            else if (down && inner.children[TOP_LEFT].type === 'inner') {
                this.render(inner.children[TOP_LEFT], topLeftNode, rowFirst, rowLast, colFirst, colLast);
            }
        }
        {
            var topRightNode = children[TOP_RIGHT];
            var down = showRight && showTop;
            if (down !== (topRightNode.dataset.node !== 'placeholder')) {
                // no match
                parent.replaceChild(down ? create(TOP_RIGHT) : placeholder(TOP_RIGHT), topRightNode);
                this.recycle(topRightNode);
            }
            else if (down && inner.children[TOP_RIGHT].type === 'inner') {
                this.render(inner.children[TOP_RIGHT], topRightNode, rowFirst, rowLast, colFirst, colLast);
            }
        }
        {
            var bottomLeftNode = children[BOTTOM_LEFT + 1];
            var down = showLeft && showBottom;
            if (down !== (bottomLeftNode.dataset.node !== 'placeholder')) {
                // no match
                parent.replaceChild(down ? create(BOTTOM_LEFT) : placeholder(BOTTOM_LEFT), bottomLeftNode);
                this.recycle(bottomLeftNode);
            }
            else if (down && inner.children[BOTTOM_LEFT].type === 'inner') {
                this.render(inner.children[BOTTOM_LEFT], bottomLeftNode, rowFirst, rowLast, colFirst, colLast);
            }
        }
        {
            var bottomRightNode = children[BOTTOM_RIGHT + 1];
            var down = showRight && showBottom;
            if (down !== (bottomRightNode.dataset.node !== 'placeholder')) {
                // no match
                parent.replaceChild(down ? create(BOTTOM_RIGHT) : placeholder(BOTTOM_RIGHT), bottomRightNode);
                this.recycle(bottomRightNode);
            }
            else if (down && inner.children[BOTTOM_RIGHT].type === 'inner') {
                this.render(inner.children[BOTTOM_RIGHT], bottomRightNode, rowFirst, rowLast, colFirst, colLast);
            }
        }
        return parent;
    };
    ACellRenderer.prototype.recycle = function (node) {
        var _this = this;
        if (node.dataset.node === 'leaf') {
            this.recycleLeaf(node);
            return;
        }
        // recycle all leaves
        var leaves = Array.from(node.querySelectorAll('[data-node=leaf]'));
        // recycle all inner nodes
        var inner = Array.from(node.querySelectorAll('[data-node=inner], [data-node=placeholder'));
        clear(node);
        leaves.forEach(function (leafNode) { return _this.recycleLeaf(leafNode); });
        inner.forEach(function (innerNode) {
            clear(innerNode);
            _this.poolInner.push(ACellRenderer.cleanUp(innerNode));
        });
        this.poolInner.push(ACellRenderer.cleanUp(node));
    };
    ACellRenderer.cleanUp = function (node) {
        if (node.style.width) {
            // eslint-disable-next-line no-param-reassign
            node.style.width = '';
        }
        if (node.style.height) {
            // eslint-disable-next-line no-param-reassign
            node.style.height = '';
        }
        return node;
    };
    ACellRenderer.prototype.recycleLeaf = function (node) {
        this.poolLeaves.push(ACellRenderer.cleanUp(node));
    };
    ACellRenderer.prototype.clearPool = function () {
        // clear pool
        this.poolInner.splice(0, this.poolInner.length);
        this.poolLeaves.splice(0, this.poolLeaves.length);
    };
    return ACellRenderer;
}());
export { ACellRenderer };
export default ACellRenderer;
